Tests/Utils/Builders/ConversationEngineTestBuilder.php

```diff
@@ -0,0 +1,398 @@
+     */
+    private function renewConversationEngine(): ConversationEngineTestBuilder
+    {
+        $this->conversationEngine = resolve(ConversationEngineInterface::class);
+
+        return $this;
+    }
+
+    /**
+     * Setting up scenario from path string
+     *
+     * @param string $path
+     * @return $this
+     */
+    private function setScenario(string $path): ConversationEngineTestBuilder
+    {
+        $scenarioFileData = File::get($path);
+        $this->scenario = ScenarioImportExportHelper::importScenarioFromString($scenarioFileData);
+
+        return $this;
+    }
+
+    /**
+     * Setting up listeners for update invents
+     *
+     * @return $this
+     */
+    private function seUpListeners(): ConversationEngineTestBuilder
+    {
+        $this->states = new Map();
+
+        resolve(Dispatcher::class)->listen(IncomingIntentStateUpdate::class, function (IncomingIntentStateUpdate $event) {
+            $this->states->put(get_class($event), $event->getConversationalState());
+        });
+
+        resolve(Dispatcher::class)->listen(OutgoingIntentStateUpdate::class, function (OutgoingIntentStateUpdate $event) {
+            $this->states->put(get_class($event), $event->getConversationalState());
+        });
+
+        resolve(Dispatcher::class)->listen(MessageTemplateStateUpdate::class, function (MessageTemplateStateUpdate $event) {
+            $this->states->put(get_class($event), $event->getConversationalState());
+        });
+
+        return $this;
+    }
+
+    /**
+     * Preparing scenario
+     *
+     * @param string $path
+     * @return $this
+     */
+    public function initializeScenario(string $path): ConversationEngineTestBuilder
+    {
+        $this->dehydrate()
+            ->renewConversationEngine()
+            ->setScenario($path)
+            ->seUpListeners();
+
+        ContextService::saveAttribute(
+            UserContext::USER_CONTEXT . '.' . 'selected_scenario',
+            $this->scenario->getUid()
+        );
+
+        return $this;
+    }
+
+    /**
+     * Set up Utterance (what user said)
+     *
+     * @param string|UtteranceAttribute $utteranceOrText
+     * @return $this
+     */
+    public function userSaid($utteranceOrText): ConversationEngineTestBuilder
+    {
+        if ($utteranceOrText instanceof UtteranceAttribute) {
+            $this->utterance = $utteranceOrText;
+        } else {
+            $this->utterance = UtteranceGenerator::generateTextUtterance($utteranceOrText);
+        }
+
+        return $this;
+    }
+
+    /**
+     * Setting up user
+     *
+     * @return $this
+     */
+    private function setUser(): ConversationEngineTestBuilder
+    {
+        $this->user = $this->utterance->getUser();
+
+        return $this;
+    }
+
+    /**
+     * Setting up user state
+     *
+     * @param ConversationalState $state
+     * @return $this
+     */
+    public function withUserState(ConversationalState $state): ConversationEngineTestBuilder
```

Comment:
What do you think about renaming this to the following?

```suggestion
    public function withConversationalState(ConversationalState $state): ConversationEngineTestBuilder
```

Reply:
hmmm... This new name has not reflected the condition of the 'existing user' in the test case. For example, 
"GIVEN Existing user with state after welcome message..."
Here we have the condition of the existing user. I understand that existing user are defined by conversation state, but from the point of intelligibility... what do you think?

Reply:
That's a fair point - I get what you mean. However, my thinking was that "with user state" potentially implies that there is some other kind of state (i.e. "with app state"). Currently there is only conversational state that is owned by a user (in user history record), however these state objects can also represent states relative to both the user and the application. I think it might be confusing for someone to see a `$state` with a speaker of `Intent::APP`, but then to see that state passed to a `withUserState` method.

Let me know what you think!